nixos/modules/services/networking/knot.nix

```diff
@@ -5,10 +5,110 @@ with lib;
+
+        # Export module sections before the template section.
+        ++ map (sec_list_fa "id" nix_def) (filter (hasPrefix "mod-") (attrNames nix_def))
+
+        ++ [ (sec_list_fa "id" nix_def "template") ]
+        ++ [ (sec_list_fa "domain" nix_def "zone") ]
+        ++ [ (sec_plain nix_def "include") ]
+      );
+
+    # A plain section contains directly attributes (we don't really check that ATM).
+    sec_plain = nix_def: sec_name: if !hasAttr sec_name nix_def then "" else
+      n2y "" { ${sec_name} = nix_def.${sec_name}; };
+
+    # This section contains a list of attribute sets.  In each of the sets
+    # there's an attribute (`fa_name`, typically "id") that must exist and come first.
+    # Alternatively we support using attribute sets instead of lists; example diff:
+    # -template = [ { id = "default"; /* other attributes */ }   { id = "foo"; } ]
+    # +template = { default = {       /* those attributes */ };  foo = { };      }
+    sec_list_fa = fa_name: nix_def: sec_name: if !hasAttr sec_name nix_def then "" else
+      let
+        elem2yaml = fa_val: other_attrs:
+          "  - " + n2y "" { ${fa_name} = fa_val; }
+          + "    " + n2y "    " other_attrs
+          + "\n";
+        sec = nix_def.${sec_name};
+      in
+        sec_name + ":\n" +
+          (if isList sec
+            then flip concatMapStrings sec
+              (elem: elem2yaml elem.${fa_name} (removeAttrs elem [ fa_name ]))
+            else concatStrings (mapAttrsToList elem2yaml sec)
+          );
+
+    # This convertor doesn't care about ordering of attributes.
+    # TODO: it could probably be simplified even more, now that it's not
+    # to be used directly, but we might want some other tweaks, too.
+    n2y = indent: val:
+      if doRecurse val then concatStringsSep "\n${indent}"
+        (mapAttrsToList
+          # This is a bit wacky - set directly under a set would start on bad indent,
+          # so we start those on a new line, but not other types of attribute values.
+          (aname: aval: "${aname}:${if doRecurse aval then "\n${indent}  " else " "}"
+            + n2y (indent + "  ") aval)
+          val
+        )
+        + "\n"
+        else
+      /*
+      if isList val && stringLength indent < 4 then concatMapStrings
+        (elem: "\n${indent}- " + n2y (indent + "  ") elem)
+        val
+        else
+      */
+      if isList val /* and long indent */ then
+        "[ " + concatMapStringsSep ", " quoteString val + " ]" else
+      if isBool val then (if val then "on" else "off") else
+      quoteString val;
+
+    # We don't want paths like ./my-zone.txt be converted to plain strings.
+    quoteString = s: ''"${if builtins.typeOf s == "path" then s else toString s}"'';
+    # We don't want to walk the insides of derivation attributes.
+    doRecurse = val: isAttrs val && !isDerivation val;
+
+  in result;
+
+  configFile = if cfg.settingsFile != null then
+    assert cfg.settings == {} && cfg.keyFiles == [];
```

Comment:
We should move this to a full fledged module system assertion to make it UX wise a lot better.
Right now I receive the following error which is a bit confusing.

```
       … while evaluating a branch condition

         at /nix/store/4h9mf2ydqjgjawihf3zkshkznm4mkavg-source/lib/customisation.nix:256:8:

          255|       drv' = (lib.head outputsList).value;
          256|     in if drv == null then null else
             |        ^
          257|       lib.deepSeq drv' drv';

       … while calling the 'head' builtin

         at /nix/store/4h9mf2ydqjgjawihf3zkshkznm4mkavg-source/lib/attrsets.nix:820:11:

          819|         || pred here (elemAt values 1) (head values) then
          820|           head values
             |           ^
          821|         else

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: assertion '(((cfg).settings == { }) && ((cfg).keyFiles == [ ]))' failed

       at /nix/store/4h9mf2ydqjgjawihf3zkshkznm4mkavg-source/nixos/modules/services/networking/knot.nix:106:5:

          105|   configFile = if cfg.settingsFile != null then
          106|     assert cfg.settings == {} && cfg.keyFiles == [];
             |     ^
          107|     cfg.settingsFile
```

Reply:
Also this isn't mentioned in the release-notes, so I am currently a bit confused how to migrate of the following:

```nix
      keyFiles = [
        config.sops.secrets."knot/keyFile".path
      ];
```

Reply:
If you switch to `settings`, you can keep `keyFiles`.

As for the module, one way is to make the assertion better, another way is to implement full compatibility.

Reply:
I ended up with this diff to have an identical knot.conf file without rewriting my config.

```diff
diff --git machines/neon/configuration.nix machines/neon/configuration.nix
index 947b152..594c898 100644
--- machines/neon/configuration.nix
+++ machines/neon/configuration.nix
@@ -134,10 +134,8 @@

     knot = {
       enable = true;
-      keyFiles = [
-        config.sops.secrets."knot/keyFile".path
-      ];
-      extraConfig = /* yaml */ ''
+      settingsFile = pkgs.writeText "knot.conf" /* yaml */ ''
+        include: ${config.sops.secrets."knot/keyFile".path}
         acl:
           - id: transfer
             key: transfer
```

Reply:
@SuperSandro2000 Thank you, this was helpful. My configuration looked very similar to yours.

Reply:
Actually, people may want to revert similar config changes when updating past PR https://github.com/NixOS/nixpkgs/pull/258361 - as such includes will surely often fail build-time validation that was disabled by accident.  (I'm sure people do want validation when *not* using secrets.)